### PR TITLE
Benchmark

### PR DESCRIPTION
**Pull request by arthur00**
This pull request will incorporate benchmarking to spacetime. Major features:
- New package called benchmark contains:
  - Packages that are test suites, with python modules that are tests
  - Benchmark and BenchmarkTest simulatinon classes used to run test suites
  - bframe: a Benchmark version of Frame that implements old-style push and pull.
  - scripts folder: contains R scripts that automatically read benchmark csv files and creates graphs

3 modes are available: BASELINE (old-style push and pull), PUSH DIFF (send modifications, retrieve everything), and PUSHPULL DIFF (diff for push and pull)

Future work: integrate Rohan's latest changes that make dependent types more efficient to pull

### CODE REVIEW

**Comment by arthur00** [comment]
Explanation of how benchmark works:

Simulation classes contain declarations like this:

```
@Producer(*BT_PRODUCER)
 @Getter(*BT_GETTER)
 @Setter(*BT_SETTER)
 @GetterSetter(*BT_GETTERSETTER)
 @Tracker(*BT_TRACKER)
 @Deleter(*BT_DELETER)
```

The variables BT_\* are global variables that are defined by the test being run. For example:

```
BT_PRODUCER =[BaseSet]
(...)
builtin_dict = globals()['__builtins__']
builtin_dict['BM_PRODUCER'] = BM_PRODUCER
```

Notice a little globals() hacking was required to get variables shared between modules, since the '@' commands are interpreted on import, not on call. Thus, a simple import test.py would automatically fill in the variables BT*. Then you can import the simulation class and those variables will exist, creating the appropriate declaration for that simulation.

The benchmark simulation classes provide optional hooks to their initialize and update loops, so each test can define their own initialize and update methods.

File: python/common/instrument.py [hunk 1]
```diff
@@ -39,26 +50,28 @@ def __init__(self, frame, filename=None):
             if hasattr(self.frame, '_instrument_headers'):
                 headers.extend(self.frame._instrument_headers)
 
-            self.fieldnames = headers
-            writer = csv.DictWriter(csvfile, delimiter=',', lineterminator='\n', fieldnames=self.fieldnames)
+            self.fieldnames = list(set(headers))
```
Comment (rohan-achar) [inline python/common/instrument.py:53]: Why list(set()) why not just list()?
It just seems to be scrambling the order?
Is that what you want to do? If yes, comment please.

File: python/benchmark/bframe.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+DIFF_PULL = set(["FULL", "DIFF_PUSHPULL"])
+DIFF_PUSH = set(["FULL", "DIFF_PUSH", "DIFF_PUSHPULL"])
+
+class BenchmarkFrame(frame):
+    def __init__(self, *args, **kwargs):
+        self.mode = None
+        super(BenchmarkFrame, self).__init__(*args, **kwargs)
+        if hasattr(self, "_instruments"):
+            inst.INSTRUMENT_HEADERS[BenchmarkFrame.__module__] = list(set(inst.INSTRUMENT_HEADERS[BenchmarkFrame.__module__]).union(set(inst.INSTRUMENT_HEADERS[frame.__module__])))  # @UndefinedVariable
```
Comment (rohan-achar) [inline python/benchmark/bframe.py:26]: same thing here.

**Review by rohan-achar — APPROVED** [review]
LGTM

**Pull request merged**